File: backend/app/api/evaluation.py
"""
GET /evaluation/{job_id} -- per knowledge-base/APPLICATION_ARCHITECTURE.md
§3. Runs evaluation.scorer.evaluate_job against the persisted job data.
"""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel

from app.api.auth import require_auth
from app.deps import AppResources, get_app_resources
from app.evaluation.scorer import evaluate_job

router = APIRouter(prefix="/evaluation", tags=["evaluation"], dependencies=[Depends(require_auth)])


class FieldAccuracyDetailResponse(BaseModel):
    row_id: str
    field_name: str
    expected: str
    actual: str | None
    matched: bool


class EvaluationResponse(BaseModel):
    job_id: str
    scored_row_count: int
    total_row_count: int
    field_level_accuracy: float | None
    char_limit_compliance: float | None
    lov_compliance: float | None
    field_accuracy_details: list[FieldAccuracyDetailResponse]
    unscored_row_ids: list[str]


@router.get("/{job_id}", response_model=EvaluationResponse)
async def get_evaluation(job_id: str, resources: AppResources = Depends(get_app_resources)) -> EvaluationResponse:
    try:
        result = await evaluate_job(resources.db_engine, job_id)
    except ValueError as exc:
        raise HTTPException(status_code=404, detail=str(exc)) from exc

    return EvaluationResponse(
        job_id=result.job_id,
        scored_row_count=result.scored_row_count,
        total_row_count=result.total_row_count,
        field_level_accuracy=result.field_level_accuracy,
        char_limit_compliance=result.char_limit_compliance,
        lov_compliance=result.lov_compliance,
        field_accuracy_details=[
            FieldAccuracyDetailResponse(
                row_id=d.row_id, field_name=d.field_name, expected=d.expected, actual=d.actual, matched=d.matched
            )
            for d in result.field_accuracy_details
        ],
        unscored_row_ids=result.unscored_row_ids,
    )

File: frontend/lib/evaluation.ts
"use client";

import { useQuery } from "@tanstack/react-query";

import { apiFetch } from "@/lib/api";
import { useApiToken, useAuthReady } from "@/lib/auth";

// Mirrors backend/app/api/evaluation.py's EvaluationResponse.
export interface FieldAccuracyDetail {
  rowId: string;
  fieldName: string;
  expected: string;
  actual: string | null;
  matched: boolean;
}

export interface Evaluation {
  jobId: string;
  scoredRowCount: number;
  totalRowCount: number;
  fieldLevelAccuracy: number | null;
  charLimitCompliance: number | null;
  lovCompliance: number | null;
  fieldAccuracyDetails: FieldAccuracyDetail[];
  unscoredRowIds: string[];
}

interface RawFieldAccuracyDetail {
  row_id: string;
  field_name: string;
  expected: string;
  actual: string | null;
  matched: boolean;
}

interface RawEvaluation {
  job_id: string;
  scored_row_count: number;
  total_row_count: number;
  field_level_accuracy: number | null;
  char_limit_compliance: number | null;
  lov_compliance: number | null;
  field_accuracy_details: RawFieldAccuracyDetail[];
  unscored_row_ids: string[];
}

export function useEvaluation(jobId: string | null) {
  const token = useApiToken();
  const authReady = useAuthReady();
  return useQuery({
    queryKey: ["evaluation", jobId],
    queryFn: async () => {
      const raw = await apiFetch<RawEvaluation>(`/evaluation/${jobId}`, undefined, token);
      const evaluation: Evaluation = {
        jobId: raw.job_id,
        scoredRowCount: raw.scored_row_count,
        totalRowCount: raw.total_row_count,
        fieldLevelAccuracy: raw.field_level_accuracy,
        charLimitCompliance: raw.char_limit_compliance,
        lovCompliance: raw.lov_compliance,
        fieldAccuracyDetails: raw.field_accuracy_details.map((d) => ({
          rowId: d.row_id,
          fieldName: d.field_name,
          expected: d.expected,
          actual: d.actual,
          matched: d.matched,
        })),
        unscoredRowIds: raw.unscored_row_ids,
      };
      return evaluation;
    },
    enabled: jobId !== null && authReady,
  });
}
